cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -0,0 +1,174 @@
+  }
+};
+
+template <typename T>
+struct scan_tile_state {
+  rmm::device_uvector<scan_tile_status> tile_status;
+  rmm::device_uvector<T> tile_state_partial;
+  rmm::device_uvector<T> tile_state_inclusive;
+
+  scan_tile_state(cudf::size_type num_tiles,
+                  rmm::cuda_stream_view stream,
+                  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+    : tile_status(rmm::device_uvector<scan_tile_status>(num_tiles, stream, mr)),
+      tile_state_partial(rmm::device_uvector<T>(num_tiles, stream, mr)),
+      tile_state_inclusive(rmm::device_uvector<T>(num_tiles, stream, mr))
+  {
+  }
+
+  operator scan_tile_state_view<T>()
+  {
+    return scan_tile_state_view<T>{tile_status.size(),
+                                   tile_status.data(),
+                                   tile_state_partial.data(),
+                                   tile_state_inclusive.data()};
+  }
+
+  inline void set_seed_async(T const seed, rmm::cuda_stream_view stream)
+  {
+    auto x = tile_status.size();
+    auto y = scan_tile_status::inclusive;
+    tile_state_inclusive.set_element_async(x - 1, seed, stream);
+    tile_status.set_element_async(x - 1, y, stream);
+  }
+
+  // T back_element(rmm::cuda_stream_view stream) const { return tile_state.back_element(stream); }
+
+  inline T get_inclusive_prefix(cudf::size_type tile_idx, rmm::cuda_stream_view stream) const
+  {
+    auto const offset = (tile_idx + tile_status.size()) % tile_status.size();
+    return tile_state_inclusive.element(offset, stream);
+  }
+};
+
+template <typename T>
+struct scan_tile_state_callback {
+  using WarpReduce = cub::WarpReduce<T>;
+
+  struct _TempStorage {
+    typename WarpReduce::TempStorage reduce;
+    T exclusive_prefix;
+  };
+
+  using TempStorage = cub::Uninitialized<_TempStorage>;
+
+  __device__ inline scan_tile_state_callback(TempStorage& temp_storage,
+                                             scan_tile_state_view<T>& tile_state,
+                                             cudf::size_type tile_idx)
+    : _temp_storage(temp_storage.Alias()), _tile_state(tile_state), _tile_idx(tile_idx)
+  {
+  }
+
+  __device__ inline T operator()(T const& block_aggregate)
+  {
+    if (threadIdx.x == 0) {
+      _tile_state.set_partial_prefix(_tile_idx, block_aggregate);  //
+    }
+
+    auto predecessor_idx    = _tile_idx - 1 - threadIdx.x;
+    auto predecessor_status = scan_tile_status::invalid;
+
+    // scan partials to form prefix
+
+    if (threadIdx.x == 0) {
+      auto window_partial = _tile_state.get_prefix(predecessor_idx, predecessor_status);
+      while (predecessor_status != scan_tile_status::inclusive) {
+        predecessor_idx--;
+        auto predecessor_prefix = _tile_state.get_prefix(predecessor_idx, predecessor_status);
+        window_partial          = predecessor_prefix + window_partial;
+      }
+      _temp_storage.exclusive_prefix = window_partial;
+    }
```

Comment:
Wouldn't we want to have this parallelised at least amongst threads of a warp. This is on the critical path of propagating the prefix forward. So we would want to minimise *latency* as much as possible. 
Seeing `WarpReduce` up there, you probably had started taking that route already :-)

Reply:
warp reduce was _significantly_ slower than this naive implementation. My guess is because of memory isn’t coalesced well with this size of state. 🤷🏻‍♂️